File: src/transform/rotation.cpp
#include "rotation.hpp"

/* RotationX */
RotationX::RotationX(const f64 _angle) {
    const f64 radian = deg2rad(_angle);
    sin_theta = std::sin(radian);
    cos_theta = std::cos(radian);
}

auto RotationX::transform_point(const p3d &p) const -> p3d {
    return p3d(
        p.x,
        cos_theta * p.y - sin_theta * p.z,
        sin_theta * p.y + cos_theta * p.z
    );
}

auto RotationX::inverse_transform_point(const p3d &p) const -> p3d {
    return p3d(
        p.x,
        sin_theta * p.z + cos_theta * p.y,
        cos_theta * p.z - sin_theta * p.y
    );
}

auto RotationX::transform_vector(const Vec3 &v) const -> Vec3 {
    return Vec3(
        v.x,
        cos_theta * v.y - sin_theta * v.z,
        sin_theta * v.y + cos_theta * v.z
    );
}

auto RotationX::inverse_transform_vector(const Vec3 &v) const -> Vec3 {
    return Vec3(
        v.x,
        sin_theta * v.z + cos_theta * v.y,
        cos_theta * v.z - sin_theta * v.y
    );
}

/* RotationY */
RotationY::RotationY(const f64 _angle) {
    const f64 radian = deg2rad(_angle);
    sin_theta = std::sin(radian);
    cos_theta = std::cos(radian);
}

auto RotationY::transform_point(const p3d &p) const -> p3d {
    return p3d(
        sin_theta * p.z + cos_theta * p.x,
        p.y,
        cos_theta * p.z - sin_theta * p.x
    );
}

auto RotationY::inverse_transform_point(const p3d &p) const -> p3d {
    return p3d(
        cos_theta * p.x - sin_theta * p.z,
        p.y,
        sin_theta * p.x + cos_theta * p.z
    );
}

auto RotationY::transform_vector(const Vec3 &v) const -> Vec3 {
    return Vec3(
        sin_theta * v.z + cos_theta * v.x,
        v.y,
        cos_theta * v.z - sin_theta * v.x
    );
}

auto RotationY::inverse_transform_vector(const Vec3 &v) const -> Vec3 {
    return Vec3(
        cos_theta * v.x - sin_theta * v.z,
        v.y,
        sin_theta * v.x + cos_theta * v.z
    );
}

/* RotationZ */
RotationZ::RotationZ(const f64 _angle) {
    const f64 radian = deg2rad(_angle);
    sin_theta = std::sin(radian);
    cos_theta = std::cos(radian);
}

auto RotationZ::transform_point(const p3d &p) const -> p3d {
    return p3d(
        cos_theta * p.x - sin_theta * p.y,
        sin_theta * p.x + cos_theta * p.y,
        p.z
    );
}

auto RotationZ::inverse_transform_point(const p3d &p) const -> p3d {
    return p3d(
        sin_theta * p.y + cos_theta * p.x,
        cos_theta * p.y - sin_theta * p.x,
        p.z
    );
}

auto RotationZ::transform_vector(const Vec3 &v) const -> Vec3 {
    return Vec3(
        cos_theta * v.x - sin_theta * v.y,
        sin_theta * v.x + cos_theta * v.y,
        v.z
    );
}

auto RotationZ::inverse_transform_vector(const Vec3 &v) const -> Vec3 {
    return Vec3(
        sin_theta * v.y + cos_theta * v.x,
        cos_theta * v.y - sin_theta * v.x,
        v.z
    );
}

File: src/transform/rotation.rs
use crate::utilities::prelude::*;
use crate::transform::prelude::*;

#[derive(Clone)]
pub struct RotationX { // Rotation around X-axis
    pub sin_theta: f64,
    pub cos_theta: f64,
}

impl RotationX {
    pub fn new(angle: f64) -> Self {
        Self {
            sin_theta: angle.to_radians().sin(),
            cos_theta: angle.to_radians().cos(),
        }
    }

    pub fn set_angle(&mut self, angle: f64) {
        self.sin_theta = angle.to_radians().sin();
        self.cos_theta = angle.to_radians().cos();
    }
}

impl Transform for RotationX {
    fn transform_point(&self, p: P3d) -> P3d {
        P3d::new(
            p.x,
            self.cos_theta * p.y - self.sin_theta * p.z,
            self.sin_theta * p.y + self.cos_theta * p.z,
        )
    }

    fn inverse_transform_point(&self, p: P3d) -> P3d {
        P3d::new(
            p.x,
            self.sin_theta * p.z + self.cos_theta * p.y,
            self.cos_theta * p.z - self.sin_theta * p.y,
        )
    }

    fn transform_vector(&self, v: Vec3) -> Vec3 {
        Vec3::new(
            v.x,
            self.cos_theta * v.y - self.sin_theta * v.z,
            self.sin_theta * v.y + self.cos_theta * v.z,
        )
    }

    fn inverse_transform_vector(&self, v: Vec3) -> Vec3 {
        Vec3::new(
            v.x,
            self.sin_theta * v.z + self.cos_theta * v.y,
            self.cos_theta * v.z - self.sin_theta * v.y,
        )
    }
}

#[derive(Clone)]
pub struct RotationY { // Rotation around Y-axis
    pub sin_theta: f64,
    pub cos_theta: f64,
}

impl RotationY {
    pub fn new(angle: f64) -> Self {
        Self {
            sin_theta: angle.to_radians().sin(),
            cos_theta: angle.to_radians().cos(),
        }
    }

    pub fn set_angle(&mut self, angle: f64) {
        self.sin_theta = angle.to_radians().sin();
        self.cos_theta = angle.to_radians().cos();
    }
}

impl Transform for RotationY {
    fn transform_point(&self, p: P3d) -> P3d {
        P3d::new(
            self.sin_theta * p.z + self.cos_theta * p.x,
            p.y,
            self.cos_theta * p.z - self.sin_theta * p.x,
        )
    }

    fn inverse_transform_point(&self, p: P3d) -> P3d {
        P3d::new(
            self.cos_theta * p.x - self.sin_theta * p.z,
            p.y,
            self.sin_theta * p.x + self.cos_theta * p.z,
        )
    }

    fn transform_vector(&self, v: Vec3) -> Vec3 {
        Vec3::new(
            self.sin_theta * v.z + self.cos_theta * v.x,
            v.y,
            self.cos_theta * v.z - self.sin_theta * v.x,
        )
    }

    fn inverse_transform_vector(&self, v: Vec3) -> Vec3 {
        Vec3::new(
            self.cos_theta * v.x - self.sin_theta * v.z,
            v.y,
            self.sin_theta * v.x + self.cos_theta * v.z,
        )
    }
}

#[derive(Clone)]
pub struct RotationZ { // Rotation around Z-axis
    pub sin_theta: f64,
    pub cos_theta: f64,
}

impl RotationZ {
    pub fn new(angle: f64) -> Self {
        Self {
            sin_theta: angle.to_radians().sin(),
            cos_theta: angle.to_radians().cos(),
        }
    }

    pub fn set_angle(&mut self, angle: f64) {
        self.sin_theta = angle.to_radians().sin();
        self.cos_theta = angle.to_radians().cos();
    }
}

impl Transform for RotationZ {
    fn transform_point(&self, p: P3d) -> P3d {
        P3d::new(
            self.cos_theta * p.x - self.sin_theta * p.y,
            self.sin_theta * p.x + self.cos_theta * p.y,
            p.z,
        )
    }

    fn inverse_transform_point(&self, p: P3d) -> P3d {
        P3d::new(
            self.sin_theta * p.y + self.cos_theta * p.x,
            self.cos_theta * p.y - self.sin_theta * p.x,
            p.z,
        )
    }

    fn transform_vector(&self, v: Vec3) -> Vec3 {
        Vec3::new(
            self.cos_theta * v.x - self.sin_theta * v.y,
            self.sin_theta * v.x + self.cos_theta * v.y,
            v.z,
        )
    }

    fn inverse_transform_vector(&self, v: Vec3) -> Vec3 {
        Vec3::new(
            self.sin_theta * v.y + self.cos_theta * v.x,
            self.cos_theta * v.y - self.sin_theta * v.x,
            v.z,
        )
    }
}
